File: src/kokoro_link/api/routes/cloud_limits.py
"""Player-facing hosted runtime-limit hints — cloud mode only.

Sibling of ``GET /cloud/credits`` and ``GET /cloud/pricing``: same
cloud-mode-only 404, same display-only contract. What it answers is "what
ceilings does this account have, and how much is spent" — the numbers the
hosted services already enforce (character slots, daily character creates,
daily 起幕 openings, per-session message cap, and the three capability
switches), so the SPA can warn *before* the player presses a button instead
of only reporting the refusal afterwards. NF4 adds the one entry that is not
a ceiling at all — ``background.dormancy_days``, the window after which the
characters stop running in the background — for the same reason: better said
in advance than discovered as an unexplained silence.

**Advisory only.** Nothing here grants or denies anything; the guards in
``character_service`` / ``story_scene_quota`` / the media services remain the
sole decision points. Accordingly the handler performs no write, and every
count is fail-soft: ``used`` may be ``null`` (with ``limit`` still present)
when a counter could not be read. A frontend that is unsure must not block —
see ``frontend/src/composables/useActionPricing.ts``.

Self-host never mounts this router (``app.py`` registers it only in cloud
mode) and the handler 404s anyway, so a self-host install's API inventory and
rendered UI are unchanged.
"""

from __future__ import annotations

import logging

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel

from kokoro_link.api.dependencies import (
    get_container,
    get_current_user,
    is_cloud_mode,
)
from kokoro_link.application.services.player_runtime_limits import (
    PlayerRuntimeLimitsSnapshot,
    QuotaUsage,
)
from kokoro_link.bootstrap.container import ServiceContainer
from kokoro_link.domain.entities.operator_profile import OperatorProfile

_LOGGER = logging.getLogger(__name__)

router = APIRouter(prefix="/cloud", tags=["cloud"])


class QuotaUsageResponse(BaseModel):
    """One ceiling. ``used`` is ``null`` when the counter was unreadable —
    the limit is still worth showing, the spent half simply is not known."""

    used: int | None
    limit: int


class BackgroundPolicyResponse(BaseModel):
    """How this account's background scheduling behaves when the player is
    away (NF4). Display-only, like everything else on this route: the
    scheduler decides, this only reports what it was told."""

    dormancy_days: int | None
    """Days without a foreground interaction after which the characters stop
    scheduling background work. ``null`` = never."""


class CloudRuntimeLimitsResponse(BaseModel):
    """A ``null`` quota means this tier does not cap that item at all."""

    character_slots: QuotaUsageResponse | None
    character_daily_creates: QuotaUsageResponse | None
    story_scenes_daily: QuotaUsageResponse | None
    session_message_limit: int | None
    album_generation_enabled: bool
    tts_enabled: bool
    video_generation_enabled: bool
    # Additive (NF4). Existing fields keep their names, types and meanings;
    # a client that has never heard of dormancy parses this body unchanged.
    background: BackgroundPolicyResponse


@router.get("/runtime-limits", response_model=CloudRuntimeLimitsResponse)
async def get_cloud_runtime_limits(
    user: OperatorProfile = Depends(get_current_user),
    container: ServiceContainer = Depends(get_container),
) -> CloudRuntimeLimitsResponse:
    if not is_cloud_mode(container):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="runtime limits are only available in cloud mode",
        )
    # Deliberately not tenant-scoped, unlike ``/cloud/credits``: the ceilings
    # come from the operator's own runtime profile, so the answer is the
    # same one the guards themselves would reach for this operator.
    service = getattr(container, "player_runtime_limits_service", None)
    if service is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="runtime limits are temporarily unavailable",
        )
    try:
        snapshot = await service.snapshot_for_operator(user.id)
    except Exception as exc:  # noqa: BLE001 - a hint must never 500 the SPA
        # The service already degrades each *counter* to ``used=None``; this
        # catches the one read it cannot degrade (the profile itself). 503,
        # not a permissive all-null body: the frontend treats 503 as "unknown"
        # and shows nothing, whereas an invented body would assert "unlimited"
        # to a player who may be one action away from a wall.
        _LOGGER.warning(
            "runtime limits snapshot failed (operator=%s)", user.id,
            exc_info=True,
        )
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="runtime limits are temporarily unavailable",
        ) from exc
    return _to_response(snapshot)


def _to_response(
    snapshot: PlayerRuntimeLimitsSnapshot,
) -> CloudRuntimeLimitsResponse:
    return CloudRuntimeLimitsResponse(
        character_slots=_quota(snapshot.character_slots),
        character_daily_creates=_quota(snapshot.character_daily_creates),
        story_scenes_daily=_quota(snapshot.story_scenes_daily),
        session_message_limit=snapshot.session_message_limit,
        album_generation_enabled=snapshot.album_generation_enabled,
        tts_enabled=snapshot.tts_enabled,
        video_generation_enabled=snapshot.video_generation_enabled,
        background=BackgroundPolicyResponse(
            dormancy_days=snapshot.background.dormancy_days,
        ),
    )


def _quota(usage: QuotaUsage | None) -> QuotaUsageResponse | None:
    if usage is None:
        return None
    return QuotaUsageResponse(used=usage.used, limit=usage.limit)

File: frontend/src/utils/api/cloudLimits.ts
/**
 * Hosted runtime limits read — cloud mode only, display-only.
 *
 * The hosted account profile (`AccountRuntimeProfile`) caps how much a player
 * may create: character slots, creations per rolling 24h, scenes per rolling
 * 24h, messages per session, plus a few feature switches. The server has
 * always enforced these; this read exists so the surfaces can *say so before*
 * the player presses the button instead of after.
 *
 * Same two failure modes as `cloudCredits`, kept distinguishable for the same
 * reason — the caller reacts differently to each:
 *
 *   - `404` — self-host (the route is not mounted at all), or a cloud operator
 *     with no projected tenant. There are no hosted limits, so every hint
 *     stays unrendered forever.
 *   - anything else — momentarily unknown. Nothing is claimed; an advisory
 *     that cannot be substantiated is simply not shown.
 *
 * Nothing here writes. A limit hint that mutated state would be a worse bug
 * than the missing hint it replaces.
 */

import { authedFetch } from '@/utils/authedFetch'

/**
 * One "N of M" counter.
 *
 * `used` is nullable on purpose: the backend serves the limit even when the
 * usage count could not be read, so the player still learns the ceiling
 * exists. A fabricated `0` would read as "plenty left" and is worse than no
 * number at all.
 */
export interface RuntimeLimitUsage {
  used: number | null
  limit: number
}

/**
 * How background scheduling behaves for this account when the player is
 * away (NF4/NF5). Not a ceiling like the fields above — it is the one entry
 * on this route that exists to be said out loud *before* it is noticed as an
 * unexplained silence, not to gate a button.
 */
export interface BackgroundPolicy {
  /**
   * Days without a foreground interaction after which characters stop
   * scheduling background work (proactive messages, schedule, world
   * events). `null` = this tier never goes dormant.
   */
  dormancyDays: number | null
}

export interface RuntimeLimitsSnapshot {
  /** Concurrent characters. `null` = uncapped. */
  character_slots: RuntimeLimitUsage | null
  /** Character creations in a rolling 24h window. `null` = uncapped. */
  character_daily_creates: RuntimeLimitUsage | null
  /** Story scenes opened in a rolling 24h window. `null` = uncapped. */
  story_scenes_daily: RuntimeLimitUsage | null
  /** Messages kept per chat session, or `null` when unbounded. */
  session_message_limit: number | null
  album_generation_enabled: boolean
  tts_enabled: boolean
  video_generation_enabled: boolean
  /**
   * `null` when the payload did not include a `background` object at all
   * (an older backend, or a malformed body) — never fabricated, so a
   * consumer that only checks `dormancyDays` still can't render on a guess.
   */
  background: BackgroundPolicy | null
}

export type RuntimeLimitsResult =
  | { kind: 'ok'; snapshot: RuntimeLimitsSnapshot }
  /** No hosted profile for this deployment / operator — never hint again. */
  | { kind: 'unsupported' }
  /** Temporarily unknown — say nothing rather than guess. */
  | { kind: 'degraded' }

const ENDPOINT = '/api/v1/cloud/runtime-limits'

export async function fetchRuntimeLimits(): Promise<RuntimeLimitsResult> {
  let res: Response
  try {
    res = await authedFetch(ENDPOINT)
  } catch {
    return { kind: 'degraded' }
  }
  if (res.status === 404) return { kind: 'unsupported' }
  if (!res.ok) return { kind: 'degraded' }
  try {
    return { kind: 'ok', snapshot: normalize(await res.json()) }
  } catch {
    return { kind: 'degraded' }
  }
}

function normalize(payload: unknown): RuntimeLimitsSnapshot {
  const body = (payload ?? {}) as Record<string, unknown>
  return {
    character_slots: usageField(body.character_slots),
    character_daily_creates: usageField(body.character_daily_creates),
    story_scenes_daily: usageField(body.story_scenes_daily),
    session_message_limit: positiveField(body.session_message_limit),
    album_generation_enabled: enabledField(body.album_generation_enabled),
    tts_enabled: enabledField(body.tts_enabled),
    video_generation_enabled: enabledField(body.video_generation_enabled),
    background: backgroundField(body.background),
  }
}

/**
 * A counter is only a counter when the ceiling is a real, non-negative
 * number. Anything else (absent, `null`, a string, a negative) means "no
 * limit to talk about" and collapses to `null` so no surface renders it.
 */
function usageField(value: unknown): RuntimeLimitUsage | null {
  if (!value || typeof value !== 'object') return null
  const entry = value as Record<string, unknown>
  const limit = entry.limit
  if (typeof limit !== 'number' || !Number.isFinite(limit) || limit < 0) {
    return null
  }
  const used = entry.used
  return {
    used: typeof used === 'number' && Number.isFinite(used) && used >= 0
      ? used
      : null,
    limit,
  }
}

function positiveField(value: unknown): number | null {
  return typeof value === 'number' && Number.isFinite(value) && value > 0
    ? value
    : null
}

/**
 * Tolerant like every other field on this route: a missing or malformed
 * `background` object (self-host on an older bundle, an unexpected shape)
 * collapses to `null` rather than throwing, so the advisory simply never
 * renders instead of crashing the read.
 */
function backgroundField(value: unknown): BackgroundPolicy | null {
  if (!value || typeof value !== 'object') return null
  const entry = value as Record<string, unknown>
  return { dormancyDays: positiveField(entry.dormancy_days) }
}

/**
 * Feature switches fail *open*: only an explicit `false` disables a feature.
 * A missing field means the backend does not model that switch, and telling
 * a player their album is off because a payload changed shape would remove a
 * feature they are entitled to. The server still refuses what it must.
 */
function enabledField(value: unknown): boolean {
  return value !== false
}
